sys/kern/kern_tc.c

```diff
@@ -1836,25 +1846,19 @@ pps_event(struct pps_state *pps, int event)
 	 * If the timecounter changed, we cannot compare the count values, so
 	 * we have to drop the rest of the PPS-stuff until the next event.
 	 */
-	if (pps->ppstc != pps->capth->th_counter) {
-		pps->ppstc = pps->capth->th_counter;
+	if (__predict_false(pps->ppstc != captc)) {
+		pps->ppstc = captc;
 		*pcount = pps->capcount;
 		pps->ppscount[2] = pps->capcount;
 		return;
 	}
 
 	/* Convert the count to a timespec. */
-	tcount = pps->capcount - pps->capth->th_offset_count;
-	tcount &= pps->capth->th_counter->tc_counter_mask;
-	bt = pps->capth->th_bintime;
-	bintime_addx(&bt, pps->capth->th_scale * tcount);
+	tcount = pps->capcount - tcount;
+	tcount &= captc->tc_counter_mask;
+	bintime_addx(&bt, capth_scale * tcount);
 	bintime2timespec(&bt, &ts);
 
-	/* If the timecounter was wound up underneath us, bail out. */
```

Comment:
This check can be removed because we saved the captured state in local variables.

Reply:
> This check can be removed because we saved the captured state in local variables.

OK. That all makes sense, and confirmed what I thought was going on, but hadn't studied the code enough to be positive about.

You've run this in a FreeBSD kernel, right?

Reply:
No, we didn't do the test runs on FreeBSD. We run the tests using the real-time operating system RTEMS. We ported the FreeBSD timecounters to RTEMS in 2015 while we added SMP support to RTEMS. During this work we also found some concurrency issues in the early implementations of the timecounters: https://github.com/freebsd/freebsd-src/commit/2c6946dca29a0481f1af751eda1ed9d97cdb92d5.

Our policy is to first integrate changes in FreeBSD and then back port them to RTEMS. We regularly synchronize the port with FreeBSD: https://github.com/RTEMS/rtems/commits/master/cpukit/score/src/kern_tc.c.

For the timecounters, we have a functional specification with validation tests: https://github.com/RTEMS/rtems-central/tree/master/spec/score/timecounter. This pull request is related to an attempt to also specify and validate the PPS synchronization. To simplify the specification it is beneficial to eliminate as many conditional expressions as possible. We also have to eliminate all dead code paths.